### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by jteo1**
Made two fixes: 
1. Mismatched curly brace detection
2. Some issues with consecutive ending curly braces when having nested directives.

Also added some number of unit tests testing basic cases of valid/invalid configs.

### CODE REVIEW

**Comment by kevinazhu** [comment]
Is the below line 45 fixed?
`//EXPECT_FALSE(parseString("server {listen 80;")); //TODO: fix this!`

Other than that, looks good.

File: config_parser.cc [hunk 1]
```diff
@@ -199,17 +201,27 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
...
       config_stack.push(new_config);
     } else if (token_type == TOKEN_TYPE_END_BLOCK) {
-      if (last_token_type != TOKEN_TYPE_STATEMENT_END) {
+      bracket_count--;
+      if (last_token_type != TOKEN_TYPE_STATEMENT_END && last_token_type != TOKEN_TYPE_END_BLOCK) { //allow two consecutive end block tockens
         // Error.
         break;
       }
+      //no matching '{' with this current '}'
+      else if(bracket_count < 0) {
```
Comment (kevinazhu) [inline config_parser.cc]: You could merge this with the statement in line 205.

File: config_parser.cc [hunk 2]
```diff
@@ -218,9 +228,16 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
     last_token_type = token_type;
   }
 
-  printf ("Bad transition from %s to %s\n",
-          TokenTypeAsString(last_token_type),
-          TokenTypeAsString(token_type));
+  if(bracket_count < 0) {
+      printf("Mismatched brackets found, extra \'}\' token(s)\n");
+  }
+  else if(bracket_count > 0) {
```
Comment (kevinazhu) [inline config_parser.cc]: If something other input breaks before the right bracket gets read, then this error will pop up, even though there is no mismatch.

File: config_parser_test.cc [hunk 3]
```diff
@@ -9,3 +9,61 @@ TEST(NginxConfigParserTest, SimpleConfig) {
...
+
+    NginxConfigParser parser;
+    NginxConfig out_config;
+};
+
+//test invalid inputs
+TEST_F(NginxStringConfigTest, InvalidConfig) {
+    EXPECT_FALSE(parseString("foo bar"));
+    EXPECT_FALSE(parseString("foo {\n}"));
+    EXPECT_FALSE(parseString("foo {\n\tx;\n};"));
```
Comment (dstephan1) [inline config_parser_test.cc]: What are these two testing 45-46? Why are they false?

Comment (jteo1) [reply config_parser_test.cc]: I'll add comments to each of these to explain as well, but I think its debatable whether foo {} is an error; I don't see any reason a programmer would want to leave it empty unless they forgot to put something in there. We can fix this later if necessary anyway, I'll just leave it as is for now.

**Pull request merged**